[DESIGN] Implement changes to bidi to permit non-strict formation (#811)
unicode-org/message-format-wg: exploration/bidi-usability.md

- U+200F RIGHT-TO-LEFT MARK (RLM)
- U+061C ARABIC LETTER MARK (ALM)

These characters are invisible strongly-directional characters.
They are used in bidirectional
text to coerce certain directional behavior (usually to mark the end of 
a sequence of characters that would otherwise be ambiguous or interact with
neutrals or opposite direction runs in an unhelpful way).

### Strictness and Abuse

We want the syntax to be somewhat permissive, particularly when it comes to paired isolates.

> The isolates and strongly-directional marks are invisble except in certain specialized editing environments.

Reviewer: ```suggestion
The isolates and strongly-directional marks are invisible except in certain specialized editing environments.
```

_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._

The syntax of a _message_ assumes a left-to-right base direction
both for the complete text of the _message_ as well as for each line (paragraph)
contained therein. 
We prefer LTR display because human understanding of a _message_ depends on LTR word tokens,
as well as token ordering (as in a placeholder or with variant keys).
Note that LTR display is **_not_** a requirement, because that is beyond the scope of MF2 itself.
However, tool and editor implementers ought to pay attention to this assumption.

Preferring LTR display is not the disadvantage to right-to-left languages that it might first appear:
- Bidi inside of _patterns_ works normally (we go to great lengths to make the interior
  of _patterns_ work as plain text)
- _Placeholders_ and _markup_ can be isolated (treated as neutrals) so that they appear
  in the correct location in an RTL _pattern_
- _Expressions_ use isolates and directional marks to display internal tokens in the
  correct order and without spillover effects
- The syntax uses paired enclosing marks that the Unicode Bidirectional Algorithm pairs

>   for shaping purposes and these offer a poor person's form of isolation.

Reviewer: I can't tell whether to read this as "paired enclosing marks (that the Unicode Bidirectional Algorithm pairs for shaping purposes)" or "uses paired enclosing marks (that the Unicode Bidirectional Algorithm pairs) for shaping purposes."

Author: It's the former.

```suggestion
- The syntax uses enclosing marks (specifically curly brackets) which the Unicode Bidirectional Algorithm
  pairs up for shaping purposes, resulting in a weak form of isolation in the syntax itself.
```

Author: Done.

We permit any of the isolate starting characters (LRI, RLI, FSI) because we want to allow
the user to set the base direction of a _literal_ or _pattern_ according to its respective 
actual contents.

> [!IMPORTANT]
> This change adds a "lookahead" to the process of determining if a given _message_ is
> "simple" or "complex", as LRI, RLI, and FSI are all valid starters for a simple message
> as well as being allowed before a quoted pattern, declaration, or selector.

This would change the ABNF as follows:
(Notice that this change includes a production `bidi` described further down
in this document)
```abnf
literal        = [open-isolate] (quoted-literal / (unquoted-literal [bidi])) [close-isolate]
quoted-pattern = [open-isolate] "{{" pattern "}}" [close-isolate]

open-isolate   = %x2066-2068
close-isolate  = %x2069
```

> [!IMPORTANT]
> The isolating characters go on the **_outside_** of the various _literal_ and _pattern_
> productions because characters on the **_inside_** of these are part of the _literal_'s
> or _pattern_'s textual content.
> We need to allow users to include bidi characters, including isolates and strongly directional marks
> in the output of MF2.

> Permit **left-to-right** isolates (`U+2066` and `U+2069`) to be used **immediately inside** the following:

Reviewer: Should this be part of a bullet list, along with the other "Permit..." imperatives that follow?

Author: The paragraph on line 331 should probably move below the paragraph "We only permit..." because it has a different meaning and refers to all isolates. Not sure about using a bullet list?

Author: Done. Created the bullet list and reorganized.

- quoted literals
- quoted patterns

We permit any of the isolate starting characters (LRI, RLI, FSI) because we want to allow
the user to set the base direction of a _literal_ or _pattern_ according to its respective 
actual contents.

This would change the ABNF as follows:
(Notice that this change includes a production `bidi` described further down
in this document)
```abnf
literal        = [open-isolate] (quoted / (unquoted [bidi])) [close-isolate]

> quoted-pattern = [open-isolate] "{{" pattern "}}" [close-isolate]

Reviewer 2: This adds LRI, RLI & FSI to the content we need to look past at message start to determine if it's complex or not. That should be specifically called out.

By contrast, if users insert too many or the wrong controls using the recommended design,
the _message_ would still be functional and would emit no undesired characters.

### Super-loose isolation

Add isolates and strongly directional marks to required and optional whitespace in the syntax.
This would permit users to get the effects described by the above design,
as long as they use isolates/marks in a "responsible" way.

(Omitting other changes found in #673)

```abnf
; strongly directional marks and bidi isolates
; ALM / LRM / RLM / LRI / RLI / FSI / PDI
bidi = %x061C / %x200E / %x200F / %x2066-2069

; optional whitespace
owsp = *( s / bidi )

; required whitespace
wsp = [ owsp ] 1*s [ owsp ]

; whitespace characters
s = ( SP / HTAB / CR / LF / %x3000 )
```

**Pros**
- Avoids problems with syntax errors that users and tools might find difficult to debug.
- Effective if used carefully.
- Addresses need to comply with UAX#31

**Cons**
- Can be used irresponsibly, including enabling some Trojan Source cases (UAX#55)

> 

Reviewer 3: ```suggestion
    - However, linters could support "responsible" use of bidi characters. (It is likely that many implementations will end up needing to support linting, such as to verify the right number and datatype of input parameters, etc.
```

Author: This would be better as a pro perhaps? "Irresponsibly" might be the wrong word in the 'con', although it smacks of code attacks, which is the idea. This design enables the right thing by also enabling many wrong things. On the other hand, a lot of the time we'll be lucky if tools implement anything or users bother to insert the isolates/marks.

Author: See most recent edits, which incorporate your suggestion into the 'con' text as well as adding more prose to the design option such that this is (hopefully) clear. I also added one more design, which tries to split-the-difference

By contrast, if users insert too many or the wrong controls using the recommended design,
the _message_ would still be functional and would emit no undesired characters.

### Super-loose isolation

Add isolates and strongly directional marks to required and optional whitespace in the syntax.
This would permit users to get the effects described by the above design,
as long as they use isolates/marks in a "responsible" way.

(Omitting other changes found in #673)

```abnf
; strongly directional marks and bidi isolates
; ALM / LRM / RLM / LRI / RLI / FSI / PDI
bidi = %x061C / %x200E / %x200F / %x2066-2069

; optional whitespace
owsp = *( s / bidi )

; required whitespace
wsp = [ owsp ] 1*s [ owsp ]

; whitespace characters
s = ( SP / HTAB / CR / LF / %x3000 )
```

**Pros**
- Avoids problems with syntax errors that users and tools might find difficult to debug.
- Effective if used carefully.
- Addresses need to comply with UAX#31

**Cons**
- Syntax does not prevent poor display outcomes, including enabling some Trojan Source cases (UAX#55);
  note that tooling or linting can help ameliorate these issues.

### Strict isolation all the time

Apply bidi isolates in a strict way.

> The main differences to the proposed solution is:

Reviewer: I only see one difference?

Author: I should do the TODO, which makes it clearer. I'll do one here in the comment for clarity and then go back and fix the PR.

The current design has `expression` thusly:

```abnf
expression = "{" [LRI] (literal-expression / variable-expression / annotation-expression) [close-isolate] "}"
```

This alternative would turn that into:

```abnf
expression = "{" (literal-expression / variable-expression / annotation-expression) "}"
           / "{" LRI (literal-expression / variable-expression / annotation-expression) close-isolate "}"
```

In this formulation, you cannot have unpaired opening (or closing) isolates without a syntax error, nor can you have multiples of open or close.

Rinse and repeat for markup, option, attribute, and literals.

Make sense?

1. The open/close isolate characters are syntactically required to be paired.
   This introduces parse errors for unpaired invisible characters,
   which could lead to bad user experiences.

As noted above, the "strict" version of the ABNF should be adopted by serializers and for 
message normalization.

// TODO put ABNF here

> Isolating rather than marking `name` helps ensure
> that its directionality does not spill over to adjoining syntax.

Reviewer 2: What follows was a part of the "Loose Isolation" alternative, parts of which are in this PR included in the proposed design; this is separate from the "Strict isolation all the time" alternative, and should have its own title as well as the ABNF that was previously included.

````suggestion

### Isolate `name` rather than `unquoted-literal`

Isolating rather than marking `name` helps ensure
that its directionality does not spill over to adjoining syntax.

The following replaces the proposed design's changes to `literal` and the `[bidi]` additions to
`variable-expression`, `function`, `option`, `attribute`, `markup`, and `ns-separator`:
```abnf
name = [open-isolate] name-start *name-char [close-isolate]
quoted-literal = [open-isolate] "|" *(quoted-char / quoted-escape) "|" [close-isolate]
```

````

Without `name` isolation, this would (misleadingly) render as
```
{⁦:אחת:שתיים⁩}
```

Note that the parsed value of the `name` does not include the open/close isolates,
just as they're not included in the parsed values of quoted literals or quoted patterns,
even though the production includes the characters.
We could accomplish this by adding an additional productions to manage `name`, at the cost
of a more complex ABNF.

**Pros**
- In the syntax, it's much simpler to include the changes to `name` in the `name` rule,
  rather than patching every place where `name` is used.

**Cons**
- Implementations need to remove isolates from the `name` token before comparing
  the value to other values (such as comparing `function` or `variable` names).
  Because of namespacing, this requires looking _inside_ the token.
- Implementations might need to insert isolates when generating names upon serialization.
  The current data model does not separate `namespace` and `name`,
  so this might be more complicated.
- `unquoted-literal` values appear as keys, as operands, and as option values.
  If not isolated, these can cause spillover effects, so we might need both `name`

>   and `unquoted-literal` isolation.

Reviewer 2: Given how `unquoted-literal` is defined as https://github.com/unicode-org/message-format-wg/blob/01f2880bd709653ad850dc746f4a72fe1133c064/spec/message.abnf#L48 and `number-literal` only contains LTR characters https://github.com/unicode-org/message-format-wg/blob/01f2880bd709653ad850dc746f4a72fe1133c064/spec/message.abnf#L50 allowing `name` isolation should cover all the `unquoted-literal` cases that may need isolation.

Author: Ur... this is something I don't like, since isolates are "not included" in the `name`, but are baked into the production. I'd rather do more with the ABNF to keep them separate (the isolates and directional characters are in the same places, but part of the surrounding goo instead of being part of the name-like token).

Note that `number-literal` contains only neutral and weakly directional characters (except for `e`, which is strongly LTR). Note that the leading `-` can switch sides in an RTL context (unless isolated or protected with LRM)
